In [160]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor

In [161]:
data = pd.read_csv("/content/Car.csv")
data.shape

(8128, 12)

In [162]:
data.isna().sum()

name               0
year               0
selling_price      0
km_driven          0
fuel               0
seller_type        0
transmission       0
owner              0
mileage          221
engine           221
max_power        216
seats            221
dtype: int64

In [163]:
data = data.dropna()
print(data.isna().sum())
print(data.shape)

name             0
year             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
seats            0
dtype: int64
(7906, 12)


In [164]:
print(data.info())
print(data.describe())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7906 entries, 0 to 8127
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           7906 non-null   object 
 1   year           7906 non-null   int64  
 2   selling_price  7906 non-null   int64  
 3   km_driven      7906 non-null   int64  
 4   fuel           7906 non-null   object 
 5   seller_type    7906 non-null   object 
 6   transmission   7906 non-null   object 
 7   owner          7906 non-null   object 
 8   mileage        7906 non-null   float64
 9   engine         7906 non-null   float64
 10  max_power      7906 non-null   float64
 11  seats          7906 non-null   float64
dtypes: float64(4), int64(3), object(5)
memory usage: 803.0+ KB
None
              year  selling_price     km_driven      mileage       engine  \
count  7906.000000   7.906000e+03  7.906000e+03  7906.000000  7906.000000   
mean   2013.983936   6.498137e+05  6.918866e+04    19.

In [165]:
check = ['name', 'fuel', 'seller_type', 'transmission', 'owner']
for c in check:
  #print(data[c].unique)
  #print(data[c].nunique)
  print(f"Value Counts for {c} is:\n {data[c].value_counts()}\n")

Value Counts for name is:
 Maruti Swift Dzire VDI                          129
Maruti Alto 800 LXI                              82
Maruti Alto LXi                                  71
BMW X4 M Sport X xDrive20d                       62
Maruti Swift VDI                                 61
                                               ... 
Honda Jazz 1.2 S AT i VTEC                        1
Hyundai i20 Active 1.2 SX with AVN                1
Ambassador CLASSIC 1500 DSL AC                    1
Tata Harrier XZ Dark Edition BSIV                 1
Toyota Innova 2.5 GX (Diesel) 8 Seater BS IV      1
Name: name, Length: 1982, dtype: int64

Value Counts for fuel is:
 Diesel    4299
Petrol    3520
CNG         52
LPG         35
Name: fuel, dtype: int64

Value Counts for seller_type is:
 Individual          6563
Dealer              1107
Trustmark Dealer     236
Name: seller_type, dtype: int64

Value Counts for transmission is:
 Manual       6865
Automatic    1041
Name: transmission, dtype: int64

V

In [166]:
data.corr()['selling_price']

<ipython-input-166-68d9344eec5b>:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()['selling_price']


year             0.412302
selling_price    1.000000
km_driven       -0.222158
mileage         -0.126280
engine           0.455682
max_power        0.749674
seats            0.041617
Name: selling_price, dtype: float64

In [167]:
data['age'] = 2023 - data['year']
data = data.drop('year', axis=1)
data

name  selling_price  km_driven    fuel  \
0           Maruti Swift Dzire VDI         450000     145500  Diesel   
1     Skoda Rapid 1.5 TDI Ambition         370000     120000  Diesel   
2         Honda City 2017-2020 EXi         158000     140000  Petrol   
3        Hyundai i20 Sportz Diesel         225000     127000  Diesel   
4           Maruti Swift VXI BSIII         130000     120000  Petrol   
...                            ...            ...        ...     ...   
8123             Hyundai i20 Magna         320000     110000  Petrol   
8124         Hyundai Verna CRDi SX         135000     119000  Diesel   
8125        Maruti Swift Dzire ZDi         382000     120000  Diesel   
8126               Tata Indigo CR4         290000      25000  Diesel   
8127               Tata Indigo CR4         290000      25000  Diesel   

     seller_type transmission                 owner  mileage  engine  \
0     Individual       Manual           First Owner    23.40  1248.0   
1     Individual       Manual          Second Owner    21.14  1498.0   
2     Individual       Manual           Third Owner    17.70  1497.0   
3     Individual       Manual           First Owner    23.00  1396.0   
4     Individual       Manual           First Owner    16.10  1298.0   
...          ...          ...                   ...      ...     ...   
8123  Individual       Manual           First Owner    18.50  1197.0   
8124  Individual       Manual  Fourth & Above Owner    16.80  1493.0   
8125  Individual       Manual           First Owner    19.30  1248.0   
8126  Individual       Manual           First Owner    23.57  1396.0   
8127  Individual       Manual           First Owner    23.57  1396.0   

      max_power  seats  age  
0         74.00    5.0    9  
1        103.52    5.0    9  
2         78.00    5.0   17  
3         90.00    5.0   13  
4         88.20    5.0   16  
...         ...    ...  ...  
8123      82.85    5.0   10  
8124     110.00    5.0   16  
8125      73.90    5.0   14  
8126      70.00    5.0   10  
8127      70.00    5.0   10  

[7906 rows x 12 columns]

In [168]:
print(data.isna().sum())

name             0
selling_price    0
km_driven        0
fuel             0
seller_type      0
transmission     0
owner            0
mileage          0
engine           0
max_power        0
seats            0
age              0
dtype: int64


In [169]:
X = data.drop('selling_price', axis=1)
y = data['selling_price']
X_trainfull, X_test, y_trainfull, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_valid, y_train, y_valid = train_test_split(X_trainfull, y_trainfull, test_size=0.2, random_state=42)
print(X_train.shape)
print(X_valid.shape)
print(X_test.shape)

(5059, 11)
(1265, 11)
(1582, 11)


In [170]:
X_test

name  km_driven    fuel seller_type  \
3756                          Maruti Swift VDI      68000  Diesel      Dealer   
5692                    Maruti Swift Dzire VDI     110000  Diesel  Individual   
3637                  Tata Indica Vista TDI LX     185000  Diesel  Individual   
7693          Toyota Corolla Altis Diesel D4DJ      90000  Diesel  Individual   
3565            Maruti Ignis 1.2 AMT Zeta BSIV      15000  Petrol  Individual   
...                                        ...        ...     ...         ...   
1359          Volkswagen Vento Diesel Highline     100000  Diesel  Individual   
641                     Mahindra Xylo E4 BS IV     168000  Diesel  Individual   
1465                 Ford Figo Diesel Titanium      50000  Diesel  Individual   
4414  Jaguar XE 2016-2019 2.0L Diesel Prestige       9000  Diesel      Dealer   
8040             Hyundai Xcent 1.2 CRDi E Plus      26000  Diesel  Individual   

     transmission                 owner  mileage  engine  max_power  seats  \
3756       Manual           First Owner    22.90  1248.0      74.00    5.0   
5692       Manual           First Owner    26.59  1248.0      74.00    5.0   
3637       Manual          Second Owner    19.40  1405.0      70.00    5.0   
7693       Manual  Fourth & Above Owner    21.43  1364.0      87.20    5.0   
3565    Automatic           First Owner    20.89  1197.0      81.80    5.0   
...           ...                   ...      ...     ...        ...    ...   
1359       Manual           Third Owner    20.54  1598.0     103.60    5.0   
641        Manual           First Owner    14.00  2498.0     112.00    7.0   
1465       Manual          Second Owner    20.00  1399.0      68.00    5.0   
4414    Automatic           First Owner    13.60  1999.0     177.00    5.0   
8040       Manual           First Owner    25.40  1186.0      73.97    5.0   

      age  
3756    9  
5692    8  
3637   11  
7693   11  
3565    5  
...   ...  
1359   10  
641    13  
1465   13  
4414    6  
8040    5  

[1582 rows x 11 columns]

In [171]:
def process_data(df):
  # One-hot Encoding
  fuel_dummies = pd.get_dummies(df['fuel'], prefix='fuel')
  seller_dummies = pd.get_dummies(df['seller_type'], prefix='seller')
  trans_dummies = pd.get_dummies(df['transmission'], prefix='trans')

  # Ordinal Encoding
  owner_order = {'Test Drive Car': 0, 'First Owner': 1, 'Second Owner': 2, 'Third Owner': 3, 'Fourth & Above Owner': 4}
  df['owner'] = df['owner'].map(owner_order)

  # Frequency Encoding
  freq = df['name'].value_counts() / len(df)
  df['name'] = df['name'].map(freq)

  # Droping Original Columns
  df = df.drop(columns=['fuel', 'seller_type', 'transmission'])

  # Concatenate One-hot Encoded Columns
  df = pd.concat([df, fuel_dummies, seller_dummies, trans_dummies], axis=1)

  # Scaling the Data
  scale = StandardScaler()
  data_scaled = scale.fit_transform(df)

  df = pd.DataFrame(data_scaled, columns=df.columns)

  return df


In [172]:
# Preprocessing the Training Data
processed_data = process_data(X_train.copy())
processed_data.shape

(5059, 17)

# Linear Regression

In [65]:
# Training Linear Regression
linear_reg = LinearRegression()
linear_reg.fit(processed_data, y_train)

LinearRegression()

In [66]:
# Predictions & Evaluation on Validation Data
y_val_pred = linear_reg.predict(process_data(X_valid))
mse_val = mean_squared_error(y_valid, y_val_pred)
rmse_val = np.sqrt(mse_val)
print("Validation Root Mean Squared Error:", rmse_val)

Validation Root Mean Squared Error: 7.488504248390997e+16


In [67]:
# Prediction & Evaluation on Test Data
y_test_pred = linear_reg.predict(process_data(X_test))
mse_test = mean_squared_error(y_test, y_test_pred)
rmse_test = np.sqrt(mse_test)
print("Test Root Mean Squared Error:", rmse_test)

Test Root Mean Squared Error: 6.966583487839733e+16


In [68]:
linear_reg.coef_

array([ 1.19050368e+05, -7.20101861e+04, -1.81250986e+04,  1.48793071e+04,
        6.35828820e+04,  4.47175490e+05, -4.07189210e+04, -1.30949836e+05,
        1.81512070e+17,  1.28792507e+18,  1.70324144e+17,  1.28531430e+18,
       -4.26580456e+17, -4.60304419e+17, -2.05038202e+17, -7.71061576e+15,
       -7.71061576e+15])

In [71]:
processed_data.columns

Index(['name', 'km_driven', 'owner', 'mileage', 'engine', 'max_power', 'seats',
       'age', 'fuel_CNG', 'fuel_Diesel', 'fuel_LPG', 'fuel_Petrol',
       'seller_Dealer', 'seller_Individual', 'seller_Trustmark Dealer',
       'trans_Automatic', 'trans_Manual'],
      dtype='object')

# Decision Tree

In [104]:
tree = DecisionTreeRegressor(random_state=42)
tree.fit(processed_data, y_train)
y_valid_pred = tree.predict(process_data(X_valid))
mse_valid = mean_squared_error(y_valid, y_valid_pred)
rmse_valid = np.sqrt(mse_valid)
print("Validation RMSE: ", rmse_valid)


Validation RMSE:  392932.7985450367


In [106]:
y_test_pred = tree.predict(process_data(X_test))
mse_test = mean_squared_error(y_test, y_test_pred)
rmse_test = np.sqrt(mse_test)
print("Test RMSE: ", rmse_test)


Test RMSE:  309072.1267422706


# Random Forest with Hyperparameters Tuning

In [125]:
from sklearn.ensemble import RandomForestRegressor

rf_reg = RandomForestRegressor(n_estimators=100, random_state=42)
rf_reg.fit(processed_data, y_train)

RandomForestRegressor(random_state=42)

In [126]:
# Predict on the validation set
y_val_pred_rf = rf_reg.predict(process_data(X_valid))

# Calculate RMSE for the validation set
mse_val_rf = mean_squared_error(y_valid, y_val_pred_rf)
rmse_val_rf = np.sqrt(mse_val_rf)
print("Validation RMSE with Random Forest:", rmse_val_rf)


Validation RMSE with Random Forest: 187970.29725546174


In [127]:
# Predict on the test set
y_test_pred_rf = rf_reg.predict(process_data(X_test))

# Calculate RMSE for the test set
mse_test_rf = mean_squared_error(y_test, y_test_pred_rf)
rmse_test_rf = np.sqrt(mse_test_rf)
print("Test RMSE with Random Forest:", rmse_test_rf)


Test RMSE with Random Forest: 195941.67334488264


## The Random Forest regressor has significantly improved the RMSE compared to the Decision Tree and Linear Regression models. The RMSE values for both validation and test sets are much closer to each other, which indicates that the model is generalizing well and not overfitting to the training data.

##Given the range of your target variable (selling_price), which spans from 29,999 to 10,000,000 with a mean around 649,813.72, an RMSE of around 187,970 for validation and 195,941 for the test is a decent result. It means, on average, the predictions are off by roughly these amounts.

In [173]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint

# Define the parameter distributions
param_distribs = {
        'n_estimators': randint(low=10, high=500),
        'max_features': ['auto', 'sqrt', 'log2'],
        'max_depth': randint(low=1, high=40),
        'min_samples_split': randint(low=2, high=40),
        'min_samples_leaf': randint(low=1, high=40),
        'bootstrap': [True, False]
    }

forest_reg = RandomForestRegressor(random_state=42)

# Use RandomizedSearchCV
rnd_search = RandomizedSearchCV(forest_reg, param_distributions=param_distribs,
                                n_iter=100, cv=5, scoring='neg_mean_squared_error',
                                random_state=42, n_jobs=-1, verbose=2)

rnd_search.fit(processed_data, y_train)

# Get the best parameters
best_params = rnd_search.best_params_
print("Best parameters found: ", best_params)

# Evaluate the model with the best hyperparameters
final_model = rnd_search.best_estimator_

y_val_pred = final_model.predict(process_data(X_valid))
mse_val = mean_squared_error(y_valid, y_val_pred)
rmse_val = np.sqrt(mse_val)
print("Validation RMSE with tuned Random Forest:", rmse_val)


Fitting 5 folds for each of 100 candidates, totalling 500 fits


/usr/local/lib/python3.10/dist-packages/sklearn/ensemble/_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(


Best parameters found:  {'bootstrap': True, 'max_depth': 12, 'max_features': 'auto', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 248}
Validation RMSE with tuned Random Forest: 173697.34941700735


In [174]:
# Predict on the test set after tuning
y_test_pred_rf = final_model.predict(process_data(X_test))

# Calculate RMSE for the test set
mse_test_rf = mean_squared_error(y_test, y_test_pred_rf)
rmse_test_rf = np.sqrt(mse_test_rf)
print("Test RMSE with Random Forest:", rmse_test_rf)


Test RMSE with Random Forest: 189499.99787524735


# Gradient Boosted Trees with Hyperparameters Tuning

In [142]:
import xgboost as xgb

# Create the XGBoost regressor
xgb_reg = xgb.XGBRegressor(objective ='reg:squarederror')

# Fit the model
xgb_reg.fit(processed_data, y_train)

# Predict on validation and test sets
y_val_pred_xgb = xgb_reg.predict(process_data(X_valid))
y_test_pred_xgb = xgb_reg.predict(process_data(X_test))

# Calculate RMSE for validation and test sets
rmse_val_xgb = np.sqrt(mean_squared_error(y_valid, y_val_pred_xgb))
rmse_test_xgb = np.sqrt(mean_squared_error(y_test, y_test_pred_xgb))

print("Validation RMSE with XGBoost:", rmse_val_xgb)
print("Test RMSE with XGBoost:", rmse_test_xgb)


Validation RMSE with XGBoost: 251062.9116908884
Test RMSE with XGBoost: 209460.91967919093


In [156]:
param_dist = {
    'n_estimators': [50, 100, 200, 300, 400],
    'learning_rate': [0.01, 0.05, 0.1, 0.2, 0.3],
    'max_depth': [3, 4, 5, 6, 7, 8],
    'colsample_bytree': [0.5, 0.7, 0.9, 1],
    'subsample': [0.5, 0.7, 0.9, 1],
    'gamma': [0, 0.1, 0.2, 0.3, 0.4],
    'min_child_weight': [1, 2, 3, 4]
}


In [157]:
from xgboost import XGBRegressor
from sklearn.model_selection import RandomizedSearchCV

xgb = XGBRegressor(objective='reg:squarederror')
random_search = RandomizedSearchCV(xgb, param_distributions=param_dist, n_iter=100, scoring='neg_mean_squared_error', n_jobs=-1, cv=5, verbose=3, random_state=42)


In [158]:
random_search.fit(processed_data, y_train)
print("Best parameters found: ", random_search.best_params_)
best_xgb_model = random_search.best_estimator_


Fitting 5 folds for each of 100 candidates, totalling 500 fits
Best parameters found:  {'subsample': 0.9, 'n_estimators': 300, 'min_child_weight': 1, 'max_depth': 3, 'learning_rate': 0.3, 'gamma': 0.4, 'colsample_bytree': 0.5}


In [159]:
y_val_pred = best_xgb_model.predict(process_data(X_valid))
mse_val = mean_squared_error(y_valid, y_val_pred)
rmse_val = np.sqrt(mse_val)
print("Validation RMSE with tuned XGBoost:", rmse_val)


Validation RMSE with tuned XGBoost: 232609.275816888
